rascal-lsp/src/main/rascal/util/LanguageServer.rsc

```diff
@@ -284,10 +301,108 @@ data DocumentSymbolTag
 
 data CompletionProposal = sourceProposal(str newText, str proposal=newText);
 
+@synopsis{Attach any command to a message for it to be exposed as a quick-fix code action automatically.}
+@description{
+The fixes you provide with a message will be hinted at by a light-bulb in the editor's margin.
+Every fix listed here will be a menu item in the pop-up menu when the bulb is activated (via short-cut or otherwise).
+}
+@benefits{
+* the information required to produce an error message is usually also required for the fix. So this
+coupling of message with fixes may come in handy.
+}
+@pitfalls{
+* the code for error messaging may become cluttered with code for fixes. It is advisable to only _collect_ information for the fix
+and store it in a ((Commmand)) constructor inside the ((CodeAction)), and not already execute the quickfix.
+}
+data Message(list[CodeAction] fixes = []);
+
+@synopsis{A Command is a parameter to a CommandExecutor function.}
+@description{
+Commands can be any closed term. Add any constructor you need to express the execution parameters
+of a command.
+
+You write the ((CommandExecutor)) to interpret each kind of ((Command)) individually.
+A ((Command) constructor must have fields or keyword fields that hold the parameters of the
+to-be-executed command. 
+
+Commands are produced for delayed and optional execution by:
+* ((LensDetector)), where the will be executed if the lens is selected in the editor 
+* ((CodeActionContributor)), where they will appear in context-menus for quick-fix and refactoring
+* ((Message)), where they will appear in context-menus on lines with error or warning diagnostics
```

Comment:
Nice! I'll add that here. Do we have a manual page on how you write TS code that does this?

Reply:
No, but #342 tracks the desire to make this easier to do.

Reply:
I'd rather not motivate people to write more TS code here. Let them stir on it and think of a Rascal solution first.
And let's add a good tutor page on #342 so that people can find the right way to do it when they think of the need.